File: lib/presentation/pages/result_page.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';
import 'package:provider/provider.dart';
import '../../core/constants/app_colors.dart';
import '../../core/constants/app_routes.dart';
import '../../core/constants/app_text_styles.dart';
import '../../core/theme/app_theme.dart';
import '../../core/utils/dummy_data.dart';
import '../../data/models/gait_result.dart';
import '../providers/session_provider.dart';
import '../widgets/app_card.dart';
import '../widgets/bottom_nav_shell.dart';
class ResultPage extends StatelessWidget {
  const ResultPage({super.key});

  @override
  Widget build(BuildContext context) {
    final provider = context.watch<SessionProvider>();
    final result = provider.currentResult ?? dummyResult;

    return Scaffold(
      backgroundColor: AppColors.background,
      body: Column(
        children: [
          _ResultAppBar(),
          Expanded(
            child: SingleChildScrollView(
              padding:
                  const EdgeInsets.all(AppTheme.marginMobile),
              child: Column(
                children: [
                  // Success header
                  Container(
                    width: 64,
                    height: 64,
                    decoration: const BoxDecoration(
                      color: AppColors.primaryContainer,
                      shape: BoxShape.circle,
                    ),
                    child: const Icon(Icons.check,
                        size: 32, color: AppColors.onPrimaryContainer),
                  ),
                  const SizedBox(height: AppTheme.md),
                  Text('Gait Analysis Result',
                      style: AppText.headlineMd,
                      textAlign: TextAlign.center),
                  const SizedBox(height: AppTheme.xl),
                  // Main score card
                  _MainScoreCard(result: result),
                  const SizedBox(height: AppTheme.md),
                  // Classification card
                  _ClassificationCard(result: result),
                  const SizedBox(height: AppTheme.md),
                  // Gait events card
                  _GaitEventsCard(result: result),
                  const SizedBox(height: AppTheme.md),
                  // Radar chart card
                  _RadarChartCard(result: result),
                  const SizedBox(height: AppTheme.xl),
                  // Action buttons
                  GestureDetector(
                    onTap: () => context.go(AppRoutes.semiogramDetail),
                    child: Container(
                      width: double.infinity,
                      padding:
                          const EdgeInsets.symmetric(vertical: 16),
                      decoration: BoxDecoration(
                        color: AppColors.primary,
                        borderRadius: AppTheme.radiusFull,
                        boxShadow: AppTheme.primaryShadow,
                      ),
                      child: Text(
                        'View Semiogram Details',
                        style: AppText.labelMd
                            .copyWith(color: AppColors.onPrimary),
                        textAlign: TextAlign.center,
                      ),
                    ),
                  ),
                  const SizedBox(height: AppTheme.sm),
                  Row(
                    children: [
                      Expanded(
                        child: _OutlineBtn(
                          label: 'Save Result',
                          icon: Icons.save_outlined,
                          onTap: () async {
                            await provider.saveSession();
                            if (context.mounted) {
                              context.go(AppRoutes.home);
                            }
                          },
                        ),
                      ),
                      const SizedBox(width: AppTheme.sm),
                      Expanded(
                        child: _OutlineBtn(
                          label: 'Export Report',
                          icon: Icons.ios_share,
                          onTap: () {},
                        ),
                      ),
                    ],
                  ),
                  const SizedBox(height: 100),
                ],
              ),
            ),
          ),
        ],
      ),
      bottomNavigationBar: const BottomNavBar(activePage: 'record'),
    );
  }
}

class _SeverityBadge extends StatelessWidget {
  final String severity;
  const _SeverityBadge({required this.severity});

  Color get _color => switch (severity.toLowerCase()) {
        'mild' => AppColors.success,
        'moderate' => AppColors.warning,
        'severe' => AppColors.error,
        _ => AppColors.onSurfaceVariant,
      };

  Color get _bg => switch (severity.toLowerCase()) {
        'mild' => AppColors.successContainer,
        'moderate' => AppColors.warningContainer,
        'severe' => AppColors.errorContainer,
        _ => AppColors.surfaceContainerHighest,
      };

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
      decoration: BoxDecoration(
        color: _bg,
        borderRadius: AppTheme.radiusFull,
      ),
      child: Text(severity,
          style: AppText.labelSm
              .copyWith(color: _color, fontWeight: FontWeight.w700)),
    );
  }
}

class _ResultAppBar extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      height: 64 + MediaQuery.of(context).padding.top,
      padding: EdgeInsets.only(
          top: MediaQuery.of(context).padding.top, left: 20, right: 20),
      color: AppColors.surface,
      child: Row(
        children: [
          GestureDetector(
            onTap: () => context.go(AppRoutes.home),
            child: Container(
              width: 40,
              height: 40,
              decoration: BoxDecoration(
                color: AppColors.surfaceContainerLow,
                borderRadius: BorderRadius.circular(12),
              ),
              child: const Icon(Icons.arrow_back,
                  size: 20, color: AppColors.onSurface),
            ),
          ),
          const SizedBox(width: 12),
          Text('Analysis Result', style: AppText.headlineSm),
        ],
      ),
    );
  }
}

class _MainScoreCard extends StatelessWidget {
  final GaitResult result;
  const _MainScoreCard({required this.result});

  @override
  Widget build(BuildContext context) {
    return AppCard(
      child: Column(
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text('FMA-LE Score', style: AppText.labelSm),
                  const SizedBox(height: 4),
                  Row(
                    crossAxisAlignment: CrossAxisAlignment.end,
                    children: [
                      Text(
                        result.fmaLeScore.toString(),
                        style: AppText.dataXl
                            .copyWith(color: AppColors.primary),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(bottom: 12),
                        child: Text(' / 34',
                            style: AppText.headlineSm.copyWith(
                                color: AppColors.onSurfaceVariant)),
                      ),
                    ],
                  ),
                ],
              ),
              _SeverityBadge(severity: result.severity),
            ],
          ),
          const SizedBox(height: AppTheme.sm),
          Row(
            children: [
              const Icon(Icons.verified_outlined,
                  size: 16, color: AppColors.success),
              const SizedBox(width: 4),
              Text(
                '${(result.confidence * 100).round()}% confidence',
                style: AppText.labelSm
                    .copyWith(color: AppColors.success),
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class _ClassificationCard extends StatelessWidget {
  final GaitResult result;
  const _ClassificationCard({required this.result});

  @override
  Widget build(BuildContext context) {
    return AppCard(
      child: Row(
        children: [
          Container(
            width: 48,
            height: 48,
            decoration: BoxDecoration(
              color: AppColors.secondaryContainer,
              borderRadius: BorderRadius.circular(12),
            ),
            child: const Icon(Icons.medical_information_outlined,
                color: AppColors.secondary),
          ),
          const SizedBox(width: AppTheme.md),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text('Classification', style: AppText.labelSm),
                Text(result.classification, style: AppText.headlineSm),
                Text(
                  'Stroke-related gait impairment pattern detected',
                  style: AppText.bodySm,
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _GaitEventsCard extends StatelessWidget {
  final GaitResult result;
  const _GaitEventsCard({required this.result});

  @override
  Widget build(BuildContext context) {
    return AppCard(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text('Gait Events', style: AppText.labelMd),
          const SizedBox(height: AppTheme.md),
          _EventRow(
            label: 'Heel Strike',
            count: result.heelStrikeCount,
            max: 40,
            color: AppColors.primary,
          ),
          const SizedBox(height: AppTheme.sm),
          _EventRow(
            label: 'Toe Off',
            count: result.toeOffCount,
            max: 40,
            color: AppColors.secondary,
          ),
        ],
      ),
    );
  }
}

class _EventRow extends StatelessWidget {
  final String label;
  final int count;
  final int max;
  final Color color;

  const _EventRow({
    required this.label,
    required this.count,
    required this.max,
    required this.color,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(label, style: AppText.labelSm),
            Text(count.toString(), style: AppText.dataViz),
          ],
        ),
        const SizedBox(height: 4),
        ClipRRect(
          borderRadius: AppTheme.radiusFull,
          child: LinearProgressIndicator(
            value: count / max,
            minHeight: 8,
            backgroundColor: AppColors.surfaceContainerHighest,
            valueColor: AlwaysStoppedAnimation<Color>(color),
          ),
        ),
      ],
    );
  }
}

class _RadarChartCard extends StatelessWidget {
  final GaitResult result;
  const _RadarChartCard({required this.result});

  @override
  Widget build(BuildContext context) {
    final s = result.semiogram;
    final values = [
      s.v,
      s.strT,
      s.ldlja,
      s.symmetryIndex,
      s.cvStrideTime,
      s.sparcrot,
    ];

    return AppCard(
      child: Column(
        children: [
          Text('Semiogram Overview', style: AppText.labelMd),
          const SizedBox(height: AppTheme.md),
          SizedBox(
            height: 220,
            child: CustomPaint(
              painter: _RadarPainter(values: values),
              child: Center(
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      '${(s.symmetryIndex * 100).round()}',
                      style: AppText.dataLg
                          .copyWith(color: AppColors.primary),
                    ),
                    Text('Index', style: AppText.labelSm),
                  ],
                ),
              ),
            ),
          ),
          const SizedBox(height: AppTheme.md),
          // Legend metrics
          Wrap(
            spacing: AppTheme.sm,
            runSpacing: AppTheme.sm,
            children: [
              _MetricChip(
                  label: 'Speed',
                  value: '${(s.v * 100).round()}%'),
              _MetricChip(
                  label: 'Springiness',
                  value: '${(s.strT * 100).round()}%'),
              _MetricChip(
                  label: 'Smoothness',
                  value: '${(s.ldlja * 100).round()}%'),
              _MetricChip(
                  label: 'Symmetry',
                  value: '${(s.symmetryIndex * 100).round()}%'),
            ],
          ),
        ],
      ),
    );
  }
}

class _MetricChip extends StatelessWidget {
  final String label;
  final String value;
  const _MetricChip({required this.label, required this.value});

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
      decoration: BoxDecoration(
        color: AppColors.surfaceContainerLow,
        borderRadius: AppTheme.radiusFull,
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text(label, style: AppText.labelSm),
          const SizedBox(width: 6),
          Text(value,
              style: AppText.dataViz
                  .copyWith(color: AppColors.primary)),
        ],
      ),
    );
  }
}

class _RadarPainter extends CustomPainter {
  final List<double> values;
  const _RadarPainter({required this.values});

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final radius = min(size.width, size.height) / 2 - 24;
    final sides = values.length;
    final angle = 2 * pi / sides;

    // Grid rings
    final gridPaint = Paint()
      ..color = AppColors.outlineVariant.withAlpha(80)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1;

    for (int ring = 1; ring <= 4; ring++) {
      final r = radius * ring / 4;
      final path = Path();
      for (int i = 0; i < sides; i++) {
        final a = -pi / 2 + i * angle;
        final x = center.dx + r * cos(a);
        final y = center.dy + r * sin(a);
        if (i == 0) path.moveTo(x, y);
        else path.lineTo(x, y);
      }
      path.close();
      canvas.drawPath(path, gridPaint);
    }

    // Axes
    for (int i = 0; i < sides; i++) {
      final a = -pi / 2 + i * angle;
      canvas.drawLine(
        center,
        Offset(center.dx + radius * cos(a), center.dy + radius * sin(a)),
        gridPaint,
      );
    }

    // Data polygon
    final dataPath = Path();
    final dataPaint = Paint()
      ..color = AppColors.primaryContainer.withAlpha(120)
      ..style = PaintingStyle.fill;
    final dataStroke = Paint()
      ..color = AppColors.primary
      ..style = PaintingStyle.stroke
      ..strokeWidth = 2;

    for (int i = 0; i < sides; i++) {
      final a = -pi / 2 + i * angle;
      final r = radius * values[i].clamp(0.0, 1.0);
      final x = center.dx + r * cos(a);
      final y = center.dy + r * sin(a);
      if (i == 0) dataPath.moveTo(x, y);
      else dataPath.lineTo(x, y);
    }
    dataPath.close();
    canvas.drawPath(dataPath, dataPaint);
    canvas.drawPath(dataPath, dataStroke);

    // Labels
    const labels = ['Speed', 'Spring', 'Smooth', 'Sym', 'Stride', 'SPARC'];
    const textStyle = TextStyle(
      fontSize: 10,
      fontFamily: 'Inter',
      color: AppColors.onSurfaceVariant,
    );
    for (int i = 0; i < sides; i++) {
      final a = -pi / 2 + i * angle;
      final lx = center.dx + (radius + 18) * cos(a);
      final ly = center.dy + (radius + 18) * sin(a);
      final tp = TextPainter(
        text: TextSpan(text: labels[i], style: textStyle),
        textDirection: TextDirection.ltr,
      )..layout();
      tp.paint(canvas,
          Offset(lx - tp.width / 2, ly - tp.height / 2));
    }
  }

  @override
  bool shouldRepaint(_RadarPainter old) => false;
}

class _OutlineBtn extends StatelessWidget {
  final String label;
  final IconData icon;
  final VoidCallback onTap;
  const _OutlineBtn(
      {required this.label, required this.icon, required this.onTap});

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(vertical: 14),
        decoration: BoxDecoration(
          borderRadius: AppTheme.radiusFull,
          border: Border.all(color: AppColors.primary, width: 1.5),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(icon, size: 18, color: AppColors.primary),
            const SizedBox(width: 6),
            Text(label,
                style:
                    AppText.labelMd.copyWith(color: AppColors.primary)),
          ],
        ),
      ),
    );
  }
}

File: src/presentation/pages/result_page.js
import { bottomNavShell, mountBottomNav } from '../widgets/bottom_nav_shell.js';
import { semiogramRadarSmall } from '../widgets/semiogram_chart.js';
import { sessionState } from '../state/session_state.js';

const PROFILE_IMG = 'https://lh3.googleusercontent.com/aida-public/AB6AXuCbM9QYN03AkFLUeD596hhW_iV9PEe2ywuDLFZ5hYa9SpE_lXNHVv7e-pI8avoOEFTNwbALsmkyyujy7d6CUHAZ9oEbHyYd_m7cvXQ8If7KYnBYbIUbAFAXHfbjW9iuasf_Xgy7cnCUQvbQdym3nky_qLIdyBoT7uAUcF33fvGiKcaBzLhuBJtRxEaM1NT04FZ4gKAYuanNAn6TYK36ytRIZQnLllPUHZo54WzN6DnKRdr5IdaMgvWo7wT-vEwqZOLnj7VLK1xziqg';

export function render() {
  const r = sessionState.currentResult;
  const confidencePct = Math.round(r.confidence * 100);
  const hsWidth = Math.round((r.heelStrikeCount / 43) * 100);
  const toWidth = Math.round((r.toeOffCount / 43) * 100);

  return `
    <div class="bg-background text-on-surface font-body-md min-h-screen pb-32">
      <!-- Top App Bar -->
      <header class="bg-surface/80 backdrop-blur-md shadow-sm sticky top-0 z-50 flex justify-between items-center px-margin-mobile h-16 w-full">
        <div class="flex items-center gap-md">
          <div class="w-10 h-10 rounded-full bg-surface-container-highest overflow-hidden border-2 border-primary-container">
            <img src="${PROFILE_IMG}" alt="Profile" class="w-full h-full object-cover">
          </div>
          <h1 class="font-headline-sm text-headline-sm text-primary">Gait Analysis</h1>
        </div>
        <div class="flex items-center gap-md">
          <button class="material-symbols-outlined text-primary transition-all duration-200 active:scale-95">sensors</button>
        </div>
      </header>

      <main class="px-margin-mobile pt-xl space-y-xl max-w-2xl mx-auto">
        <!-- Success Header -->
        <div class="text-center">
          <div class="inline-flex items-center justify-center w-12 h-12 rounded-full bg-primary-container text-on-primary-container mb-sm">
            <span class="material-symbols-outlined" style="font-variation-settings: 'FILL' 1;">check_circle</span>
          </div>
          <h2 class="font-headline-md text-headline-md text-on-surface">Gait Analysis Result</h2>
          <p class="font-body-md text-on-surface-variant">Session completed successfully</p>
        </div>

        <!-- Score + Cards Bento Grid -->
        <div class="grid grid-cols-1 md:grid-cols-2 gap-md">
          <!-- Primary Score -->
          <div class="md:col-span-2 bg-surface-container-lowest rounded-lg p-lg shadow-[0_4px_20px_0_rgba(29,53,87,0.06)] flex flex-col items-center justify-center relative overflow-hidden">
            <div class="absolute top-0 right-0 p-md">
              <span class="bg-tertiary-container text-on-tertiary-container px-sm py-1 rounded-full text-label-sm font-label-md">${r.severity}</span>
            </div>
            <div class="text-center space-y-xs">
              <p class="font-label-md text-on-surface-variant uppercase tracking-wider">Overall Assessment</p>
              <div class="flex items-baseline justify-center gap-xs">
                <span class="font-display-lg text-[64px] text-primary">${r.fmaLeScore}</span>
                <span class="font-headline-sm text-on-surface-variant">/ 34</span>
              </div>
              <p class="font-headline-sm text-on-surface">FMA-LE Score</p>
              <div class="flex items-center justify-center gap-xs text-primary-container font-label-md">
                <span class="material-symbols-outlined text-[18px]" style="font-variation-settings: 'FILL' 1;">verified</span>
                <span>${confidencePct}% confidence</span>
              </div>
            </div>
          </div>

          <!-- Classification -->
          <div class="bg-surface-container-lowest rounded-lg p-lg shadow-[0_4px_20px_0_rgba(29,53,87,0.06)] border-l-4 border-primary">
            <div class="flex items-center gap-sm mb-base">
              <span class="material-symbols-outlined text-primary">clinical_notes</span>
              <h3 class="font-label-md text-on-surface-variant">Classification</h3>
            </div>
            <p class="font-display-lg text-primary">${r.classification}</p>
            <p class="font-body-md text-on-surface-variant mt-xs">Cerebrovascular Accident pattern detected in gait cycle.</p>
          </div>

          <!-- Gait Events -->
          <div class="bg-surface-container-lowest rounded-lg p-lg shadow-[0_4px_20px_0_rgba(29,53,87,0.06)]">
            <div class="flex items-center gap-sm mb-base">
              <span class="material-symbols-outlined text-primary">timeline</span>
              <h3 class="font-label-md text-on-surface-variant">Gait Events</h3>
            </div>
            <div class="space-y-sm">
              <div class="flex justify-between items-center">
                <span class="font-body-md text-on-surface">Heel Strike</span>
                <span class="font-data-viz text-primary text-headline-sm">${r.heelStrikeCount}</span>
              </div>
              <div class="w-full bg-surface-container rounded-full h-1.5">
                <div class="bg-primary h-1.5 rounded-full" style="width:${hsWidth}%"></div>
              </div>
              <div class="flex justify-between items-center">
                <span class="font-body-md text-on-surface">Toe Off</span>
                <span class="font-data-viz text-primary text-headline-sm">${r.toeOffCount}</span>
              </div>
              <div class="w-full bg-surface-container rounded-full h-1.5">
                <div class="bg-primary-container h-1.5 rounded-full" style="width:${toWidth}%"></div>
              </div>
            </div>
          </div>

          <!-- Semiogram Summary -->
          <div class="md:col-span-2 bg-surface-container-lowest rounded-lg p-lg shadow-[0_4px_20px_0_rgba(29,53,87,0.06)] overflow-hidden">
            <div class="flex justify-between items-center mb-lg">
              <div class="flex items-center gap-sm">
                <span class="material-symbols-outlined text-primary">radar</span>
                <h3 class="font-label-md text-on-surface-variant">Semiogram Summary</h3>
              </div>
              <span class="text-label-sm font-label-md text-primary bg-primary-fixed/30 px-sm py-1 rounded-full uppercase tracking-tighter">Live Analysis</span>
            </div>
            <div class="flex flex-col md:flex-row items-center gap-xl">
              ${semiogramRadarSmall()}
              <div class="flex-1 grid grid-cols-2 gap-md w-full">
                <div class="p-sm bg-surface rounded-lg">
                  <p class="font-label-sm text-on-surface-variant">Speed</p>
                  <p class="font-data-viz text-headline-sm text-on-surface">88%</p>
                </div>
                <div class="p-sm bg-surface rounded-lg">
                  <p class="font-label-sm text-on-surface-variant">Stability</p>
                  <p class="font-data-viz text-headline-sm text-on-surface">64%</p>
                </div>
                <div class="p-sm bg-surface rounded-lg">
                  <p class="font-label-sm text-on-surface-variant">Symmetry</p>
                  <p class="font-data-viz text-headline-sm text-on-surface">42%</p>
                </div>
                <div class="p-sm bg-surface rounded-lg">
                  <p class="font-label-sm text-on-surface-variant">Smoothness</p>
                  <p class="font-data-viz text-headline-sm text-on-surface">76%</p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <!-- Action Buttons -->
        <div class="space-y-md pt-base">
          <button id="view-semiogram-btn" class="w-full bg-primary text-on-primary py-md px-lg rounded-full font-headline-sm shadow-md transition-all duration-200 active:scale-95 hover:opacity-90 flex justify-center items-center gap-sm">
            View Semiogram Details
            <span class="material-symbols-outlined">arrow_forward</span>
          </button>
          <div class="grid grid-cols-2 gap-md">
            <button class="w-full bg-secondary-container text-on-secondary-container py-md px-base rounded-full font-label-md flex items-center justify-center gap-xs transition-all duration-200 active:scale-95">
              <span class="material-symbols-outlined">save</span> Save Result
            </button>
            <button id="new-test-btn" class="w-full bg-secondary-container text-on-secondary-container py-md px-base rounded-full font-label-md flex items-center justify-center gap-xs transition-all duration-200 active:scale-95">
              <span class="material-symbols-outlined">refresh</span> New Test
            </button>
          </div>
        </div>
      </main>

      ${bottomNavShell('record')}
    </div>
  `;
}

export function mount(navigate) {
  mountBottomNav(navigate);
  document.getElementById('view-semiogram-btn')?.addEventListener('click', () => navigate('/semiogram-detail'));
  document.getElementById('new-test-btn')?.addEventListener('click', () => navigate('/instruction'));
}
